File: analytics/src/test/java/io/appmetrica/analytics/impl/crash/jvm/JvmCrashReaderTest.java
package io.appmetrica.analytics.impl.crash.jvm;

import android.os.ResultReceiver;
import io.appmetrica.analytics.TestData;
import io.appmetrica.analytics.impl.ClientCounterReport;
import io.appmetrica.analytics.impl.CounterReport;
import io.appmetrica.analytics.impl.IOUtils;
import io.appmetrica.analytics.impl.client.ClientConfiguration;
import io.appmetrica.analytics.impl.client.ProcessConfiguration;
import io.appmetrica.analytics.internal.CounterConfiguration;
import io.appmetrica.analytics.testutils.CommonTest;
import java.io.File;
import java.util.HashMap;
import org.json.JSONException;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class JvmCrashReaderTest extends CommonTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    private File crashFile;
    @InjectMocks
    private JvmCrashReader crashReader;

    @Test
    public void hasData() throws JSONException {
        try (MockedStatic<IOUtils> sIoUtils = Mockito.mockStatic(IOUtils.class)) {
            CounterReport counterReport = new CounterReport();
            counterReport.setName("crash");
            String crashValue = new JvmCrash(
                    counterReport,
                    new ClientConfiguration(new ProcessConfiguration(
                            RuntimeEnvironment.getApplication(),
                            mock(ResultReceiver.class)
                    ), new CounterConfiguration(TestData.TEST_UUID)),
                    new HashMap<ClientCounterReport.TrimmedField, Integer>()
            ).toJSONString();
            when(IOUtils.getStringFileLocked(crashFile)).thenReturn(crashValue);
            assertThat(crashReader.apply(crashFile).toJSONString()).isEqualTo(crashValue);
        }
    }

    @Test
    public void noData() {
        try (MockedStatic<IOUtils> sIoUtils = Mockito.mockStatic(IOUtils.class)) {
            when(IOUtils.getStringFileLocked(crashFile)).thenReturn(null);
            assertThat(crashReader.apply(crashFile)).isNull();
        }
    }

    @Test
    public void emptyData() {
        try (MockedStatic<IOUtils> sIoUtils = Mockito.mockStatic(IOUtils.class)) {
            when(IOUtils.getStringFileLocked(crashFile)).thenReturn("");
            assertThat(crashReader.apply(crashFile)).isNull();
        }
    }

    @Test
    public void deleteFile() {
        crashReader.consume(crashFile);
        verify(crashFile).delete();
    }
}

File: analytics/src/testDebug/java/io/appmetrica/analytics/impl/crash/jvm/JvmCrashReaderTest.kt
package io.appmetrica.analytics.impl.crash.jvm

import io.appmetrica.analytics.impl.IOUtils
import io.appmetrica.gradle.testutils.CommonTest
import io.appmetrica.gradle.testutils.rules.MockedConstructionRule.Companion.constructionRule
import io.appmetrica.gradle.testutils.rules.MockedStaticRule.Companion.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.io.File

internal class JvmCrashReaderTest : CommonTest() {

    private val fileModifiedTimestamp = 1700000000000L

    private val crashFile: File = mock {
        on { lastModified() } doReturn fileModifiedTimestamp
    }

    @get:Rule
    val ioUtilsRule = staticRule<IOUtils>()

    @get:Rule
    val jvmCrashRule = constructionRule<JvmCrash>()

    private val crashReader: JvmCrashReader by setUp { JvmCrashReader() }

    @Test
    fun handleCrashDataReturnsNullForBlankData() {
        assertThat(crashReader.handleCrashData(crashFile, null, fileModifiedTimestamp)).isNull()
        assertThat(crashReader.handleCrashData(crashFile, "", fileModifiedTimestamp)).isNull()
    }

    @Test
    fun handleCrashDataCreatesAndReturnsJvmCrashWithCorrectArgs() {
        val crashData = "some crash json"
        val result = crashReader.handleCrashData(crashFile, crashData, fileModifiedTimestamp)
        assertThat(jvmCrashRule.constructionMock.constructed()).hasSize(1)
        assertThat(result).isSameAs(jvmCrashRule.constructionMock.constructed()[0])
        assertThat(jvmCrashRule.argumentInterceptor.flatArguments())
            .containsExactly(crashData, fileModifiedTimestamp)
    }

    @Test
    fun applyReturnsNullWhenIOUtilsReturnsNull() {
        whenever(IOUtils.getStringFileLocked(crashFile)).thenReturn(null)
        assertThat(crashReader.apply(crashFile)).isNull()
    }

    @Test
    fun applyCreatesAndReturnsJvmCrashForValidData() {
        val crashData = "some crash json"
        whenever(IOUtils.getStringFileLocked(crashFile)).thenReturn(crashData)
        val result = crashReader.apply(crashFile)

        assertThat(jvmCrashRule.constructionMock.constructed()).hasSize(1)
        assertThat(result).isSameAs(jvmCrashRule.constructionMock.constructed()[0])
        assertThat(jvmCrashRule.argumentInterceptor.flatArguments())
            .containsExactly(crashData, fileModifiedTimestamp)
    }

    @Test
    fun consumeDeletesCrashFile() {
        whenever(crashFile.delete()).thenReturn(true)
        crashReader.consume(crashFile)
        verify(crashFile).delete()
    }

    @Test
    fun consumeHandlesFalseDeleteWithoutThrowing() {
        whenever(crashFile.delete()).thenReturn(false)
        crashReader.consume(crashFile)
        verify(crashFile).delete()
    }

    @Test
    fun consumeHandlesExceptionFromDeleteWithoutThrowing() {
        whenever(crashFile.delete()).thenThrow(SecurityException("no permission"))
        crashReader.consume(crashFile)
        verify(crashFile).delete()
    }
}
